### ISSUE TITLE
Add tags to project titles?

### ISSUE THREAD

**Issue by samael205** [fabiospampinato/vscode-todo-plus#58]
Hi fabio.

I want to add tags to te titles of my projects.

A  project looks like this
![proyecto normal](https://user-images.githubusercontent.com/649740/40477928-ff5f31a2-5f47-11e8-8a9e-f7df7c8174b6.jpg)

A project with a task with a tag looks like this
![proyecto normal con tag](https://user-images.githubusercontent.com/649740/40478107-7cc8899a-5f48-11e8-8f7c-3f56f52211f7.jpg)

And a project with tag with a task with tag looks like this
![proyecto con tag normal con tag](https://user-images.githubusercontent.com/649740/40478269-e4cf6392-5f48-11e8-980f-7cc2ba944cb2.jpg)

Right now for every tag that I add to the title of the project, I see the calculations of the missing tasks on the right

Is it easy to solve or is it possible to solve it?

The version of Todo+ is 1.18.2

**Comment by fabiospampinato** [comment]
Project-level statistics shouldn't get duplicated, I'll take a look into that.

I'm currently refactoring the repo in a major way, so let me first do that, I'll probably push an update about this today 👍

**Comment by fabiospampinato** [comment]
It will, in some scenarios the number of lines in the document won't matter.

Anyway project-level statistics are still pretty heavy on the CPU, I'll probably disable them by default until I can find a solution to this.

**Comment by samael205** [comment]
Ok, but, I just realized that I have disabled a couple of extensions and now it is much better. I'm going to try another file that has more than 1500 lines

**Comment by fabiospampinato** [comment]
I'm currently experimenting with a 7k+ lines file, and the situation seems much better than it is right now.

**Comment by samael205** [comment]
I've been disabling a lot of extensions, and I've noticed that every time I write fast the CPU goes up to 45% - 50%. I do not know if it's Visual Studio Code or my CPU problem. It's not your extension problem.

**Comment by fabiospampinato** [comment]
If that happens also on other kind of files than it's not caused by Todo+, at least in those cases.

**Comment by fabiospampinato** [comment]
I've just published v2.0.0. 

You should add tags to projects like this:

```
Project: @tag
```

Rather then:

```
Project @tag :
```

I'm keeping this open for now as it's still kind of buggy sometimes.

**Comment by samael205** [comment]
Wow, my congratulations, improved a lot!

**Comment by samael205** [comment]
Is this behavior normal?

I have the project prinicpal project
that project is divided into a task and two projects that have other tasks
The main project task will cost me a time of 1 minute.

the first task of the son project costs me a time of 1 hour

Within that project, son, I have another project, which has a task that will take 2 days and another one that will take 1h

I think the total project time itself is 2d2h1m, right?

![strangething](https://user-images.githubusercontent.com/649740/40531438-ba9b0982-5ffc-11e8-86e7-0307c213b4cc.jpg)

**Comment by samael205** [comment]
`"todo.statistics.project.text": "([all]) / ([pending]) - ([percentage])% - [est]"`

**Comment by fabiospampinato** [comment]
Can you also post your todo file as text?

**Comment by samael205** [comment]
`Project:
    ☐ Create new project @1m @created(18-05-25 09:05)
    sub-project:
        ☐ Create a subProject @1h @created(18-05-25 09:05)
        ☐ Create another task @created(18-05-25 09:05)
        ☐ a task @created(18-05-25 09:06)
            sub-sub-project:
                ☐ again a task @2d @created(18-05-25 09:06)
                ☐ another project @1h @created(18-05-25 09:06)`

**Comment by samael205** [comment]
[prueba.zip](https://github.com/fabiospampinato/vscode-todo-plus/files/2038842/prueba.zip)
The file

**Comment by fabiospampinato** [comment]
Fixed both problems in v2.0.1.

**Comment by samael205** [comment]
I'm sure you'll want to kill me, but I've discovered something new.

I guess the indentation establishes the projects that there are,
and that if there are two projects on the same indentation level, they are treated as different projects.

What happens now is that the hours of two different projects are added, when not indentation


[prueba.zip](https://github.com/fabiospampinato/vscode-todo-plus/files/2039096/prueba.zip)


Try again this file, indent the second project

**Comment by fabiospampinato** [comment]
On the contrary, the more bugs you can find the better I can make this extension :)

Let me take a look at that!

**Comment by fabiospampinato** [comment]
I made a typo https://github.com/fabiospampinato/vscode-todo-plus/commit/3ad39f09aacf1800889230f31e336bae33609b0a :D 

It should fixed now in v2.0.2. Let me know if you find anything else.

**Comment by samael205** [comment]
Works perfect @fabiospampinato.

Thanks!, i will post a new idea for enchancement the extension. I think that i can do a pull Request, but im not sure.

**Comment by samael205** [comment]
@fabiospampinato The problem of the high cost of cpu is reproduced well in this file, and it only happens when you have to apply styles in between and you are writing.
For example, when creating a new task, it is created such that:

☐ @created (05-18-28 11:37)

This task already has some defined styles, so every time you type something, the CPU grows up to 45-50%

This only happens with the files that handle the extension

There are times when it is very frustrating to write, because I have to wait 30 seconds for what I write to be reflected on the screen

I leave the sample file to see if you can play it. This in Spanish
[todoExample.zip](https://github.com/fabiospampinato/vscode-todo-plus/files/2044570/todoExample.zip)

**Comment by fabiospampinato** [comment]
@samael205 you can read more about my recent performance optimizations [here](https://github.com/fabiospampinato/vscode-todo-plus/issues/54). 

Unfortunately the case you mentioned isn't one of the `O(1)` cases, but maybe I could optimize that one as well, I'll have to think about it 🤔 

Those files are quite big and the extension is redecorating the entire document over and over again, hence the slowness.

Waiting 30s is definitely too much. Are you running the extension on not-so up-to-date hardware?

Anyway I can't optimize the extension properly for all cases because there are some necessary APIs missing, but maybe I'll find a way to optimize your scenario in some other way.

**Comment by samael205** [comment]
Would not the following be a solution?
You load the file, you go line to line and apply the styles that are necessary.
Once this is done, you simply have to rewrite the styles of the line where the cursor is at that moment.

I say it for giving some idea

**Comment by fabiospampinato** [comment]
@samael205 that's pretty much exactly what I would like to do, but the APIs needed for this do not exists: https://github.com/Microsoft/vscode/issues/50346
